File: Python/models/TaskTimer.py
from utils import utility
import time


class TaskTimer:
    def __init__(self, name):
        self.name = name
        self.uid = utility.make_id(6)
        self.updated_at = time.time()
        self.action_list = [{"action": "created", "date": self.updated_at}]

    def start(self):
        self.updated_at = time.time()
        self.action_list.append({"action": "started", "date": self.updated_at})

    def pause(self):
        self.updated_at = time.time()
        self.action_list.append({"action": "paused", "date": self.updated_at})

    def reset(self):
        self.updated_at = time.time()
        self.action_list = [{"action": "created", "date": self.updated_at}]

    def get_status(self):
        return self.action_list[-1]["action"]

    def get_elapsed_time(self):
        last_action = self.action_list[-1]
        elapsed_time = time.time() - last_action["date"]
        return elapsed_time

    def get_uid(self):
        return self.uid


class AllTimers:
    _instance = None

    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super(AllTimers, cls).__new__(cls)
            cls._instance.timer_list = {}
        return cls._instance

    def create_timer(self, name):
        if name in self.timer_list:
            raise ValueError(f"Timer with name \"{name}\" already exists.")
        self.timer_list[name] = TaskTimer(name)

    def start_timer(self, name):
        if name not in self.timer_list:
            raise ValueError(f"Timer with name \"{name}\" does not exist.")
        if self.timer_list[name].get_status() == "started":
            raise ValueError("Timer already started.")
        elif self.timer_list[name].get_status() == "stopped":
            raise ValueError("Timer is stopped.")
        self.timer_list[name].start()

    def pause_timer(self, name):
        if name not in self.timer_list:
            raise ValueError(f"Timer with name \"{name}\" does not exist.")
        if self.timer_list[name].get_status() == "paused":
            raise ValueError("Timer already paused.")
        elif self.timer_list[name].get_status() == "stopped":
            raise ValueError("Timer is stopped.")
        self.timer_list[name].pause()

    def reset_timer(self, name):
        if name not in self.timer_list:
            raise ValueError(f"Timer with name \"{name}\" does not exist.")
        self.timer_list[name].reset()

    def stop_timer(self, name):
        if name not in self.timer_list:
            raise ValueError(f"Timer with name \"{name}\" does not exist.")
        if self.timer_list[name].get_status() == "stopped":
            raise ValueError("Timer already stopped.")
        self.timer_list[name].stop()

    def get_status_timer(self, name):
        if name not in self.timer_list:
            raise ValueError(f"Timer with name \"{name}\" does not exist.")
        return self.timer_list[name].get_status()

    def get_elapsed_time_timer(self, name):
        if name not in self.timer_list:
            raise ValueError(f"Timer with name \"{name}\" does not exist.")
        return self.timer_list[name].get_elapsed_time()

    def get_uid_timer(self, name):
        if name not in self.timer_list:
            raise ValueError(f"Timer with name \"{name}\" does not exist.")
        return self.timer_list[name].get_uid()

    def show_all_timers(self):
        return [f" {key}" for key in self.timer_list.keys()]

File: C#/Models/TaskTimer.cs
using System;

namespace CLI_Task_Timer_CS.Models
{
    public class TaskTimer
    {
        private string Name { get; set; }
        private string Uid { get; set; }
        private DateTime UpdatedAt { get; set; }
        private List<(string action, DateTime date)> ActionList { get; set; }

        public TaskTimer(string name)
        { 
            var utility = new Utility();
            Name = name;
            Uid = utility.makeID(6);
            UpdatedAt = DateTime.Now;
            ActionList = new List<(string action, DateTime date)>();
            ActionList.Add(("created", DateTime.Now));
        }


        public void start()
        {
            UpdatedAt = DateTime.Now;
            ActionList.Add(("started", DateTime.Now));
        }
        public void stop()
        {
            UpdatedAt = DateTime.Now;
            ActionList.Add(("stopped", DateTime.Now));
        }
        public void pause()
        {
            UpdatedAt = DateTime.Now;
            ActionList.Add(("paused", DateTime.Now));
        }
        public void reset ()
        {
            UpdatedAt = DateTime.Now;
            ActionList.Add(("created", DateTime.Now));
        }
        public string getStatus()
        {
            if (ActionList.Count == 0)
            {
                return "No actions recorded.";
            }
            var latestAction = ActionList[ActionList.Count - 1];
            return $"Last action: {latestAction.action} at {latestAction.date}";
        }
        public string getElapsedTime()
        {
            return ("Time since task " + ActionList[ActionList.Count - 1].action +
                    ActionList[ActionList.Count - 1].date);
        }
        public string getUid()
        {
            return Uid;
        }


}
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
}